src/mappings/HistoryElements.ts

```diff
@@ -149,38 +156,79 @@ function findFailedTransferCalls(extrinsic: SubstrateExtrinsic): Array<TransferD
             (transfer as AssetTransfer).assetId = assetId
         }
 
-        return {
+        return [{
             isTransferAll,
             transfer,
+        }]
+    }
+
+    let assetIdFee = "native"
+    let fee = calculateFeeAsString(extrinsic)
+    let foundAssetTxFeePaid = extrinsic.block.events.find((e) => isAssetTxFeePaidEvent(eventRecordToSubstrateEvent(e)));
+    if (foundAssetTxFeePaid !== undefined)  {
+        const [who, actual_fee, tip, rawAssetIdFee] = getEventData(eventRecordToSubstrateEvent(foundAssetTxFeePaid))
+        if ('interior' in rawAssetIdFee) {
+            assetIdFee = getAssetIdFromMultilocation(rawAssetIdFee)
+            fee = actual_fee.toString();
         }
-    })
+    }
+    const swapCallback = (path, amountIn, amountOut, receiver) => {
+        const swap: Swap = {
+            assetIdIn: getAssetIdFromMultilocation(path[0]),
+            amountIn: amountIn.toString(),
+            assetIdOut: getAssetIdFromMultilocation(path[path["length"] - 1]),
+            amountOut: amountOut.toString(),
+            sender: sender.toString(),
+            receiver: receiver.toString(),
+            assetIdFee: assetIdFee,
+            fee: fee,
+            eventIdx: -1,
+            success: false
+        }
+
+        return [{
+            isTransferAll: false,
+            transfer: swap,
+        }]
+    }
+
+    let transferCalls = determineTransferCallsArgs(extrinsic.extrinsic.method, transferCallback, swapCallback)
+    if (transferCalls.length == 0) {
+        return null;
+    }
+
+    return transferCalls
 }
 
-function determineTransferCallsArgs(causeCall: CallBase<AnyTuple>) : [boolean, string, bigint, string?][] {
+function determineTransferCallsArgs(causeCall: CallBase<AnyTuple>, transferCallback, swapCallback) : Array<TransferData> {
     if (isNativeTransfer(causeCall)) {
```

Comment:
Looks like our visitor approach is not implemented correctly. The initial idea was to pass a compound visitor - single closure (that might contain a list of subvisitors). Then on passes an object to the visitor which iterates subvisitors to try to transform the object and returns first not null result (for example).